src/imageunpacker/serviceimage.cpp

```diff
@@ -0,0 +1,274 @@
+
+    if (auto err = aos::common::utils::UnpackTarImage(archivePath, value); !err.mError.IsNone()) {
+        throw std::runtime_error("Failed to unpack image: " + err.mValue);
+    }
+
+    return value;
+}
+
+static void ValidateImage(const std::string& imagePath, const ServiceManifest& serviceManifest)
+{
+    LOG_DBG() << "Validating image: " << imagePath.c_str();
+
+    if (auto res = aos::common::utils::ValidateDigest(serviceManifest.mManifest.mConfig.mDigest);
+        !res.mError.IsNone()) {
+        throw std::runtime_error("Invalid image config digest: " + res.mValue);
+    }
+
+    if (serviceManifest.mAosService.has_value()) {
+        if (auto res = aos::common::utils::ValidateDigest(serviceManifest.mAosService->mDigest); !res.mError.IsNone()) {
+            throw std::runtime_error("Invalid aos service digest: " + res.mValue);
+        }
+
+        auto [algorithm, hex] = aos::common::utils::ParseDigest(serviceManifest.mAosService->mDigest);
+
+        // this is only for validate json
+        ParseJson(fs::path(imagePath) / kBlobsFolder / algorithm / hex);
+    }
+
+    auto [algorithm, hex] = aos::common::utils::ParseDigest(serviceManifest.mManifest.mLayers[0].mDigest);
+    auto rootfsPath       = fs::path(imagePath) / kBlobsFolder / algorithm / hex;
+
+    if (!fs::exists(rootfsPath)) {
+        throw std::runtime_error("Rootfs not found: " + rootfsPath.string());
+    }
+
+    if (!fs::is_directory(rootfsPath)) {
+        if (auto res = aos::common::utils::ValidateDigest(serviceManifest.mManifest.mLayers[0].mDigest);
+            !res.mError.IsNone()) {
+            throw std::runtime_error("Invalid rootfs digest: " + res.mValue);
+        }
+
+        return;
+    }
+
+    auto [digest, error] = aos::common::utils::HashDir(rootfsPath);
+    if (!error.IsNone()) {
+        throw std::runtime_error("Failed to calculate rootfs checksum: " + digest);
+    }
+
+    if (serviceManifest.mManifest.mLayers[0].mDigest != digest) {
+        throw std::runtime_error("Incorrect rootfs checksum");
+    }
+}
+
+static aos::common::utils::Digest PrepareServiceFS(const std::string& imagePath, const ServiceManifest& serviceManifest)
+{
+    LOG_DBG() << "Preparing service FS: " << imagePath.c_str();
+
+    auto imageParts = GetImageParts(imagePath, serviceManifest);
+
+    auto tmpRootFS = fs::path(imagePath) / kTmpRootFSDir;
+    fs::create_directory(tmpRootFS);
+
+    if (auto err = aos::common::utils::UnpackTarImage(imageParts.mServiceFSPath, tmpRootFS); !err.mError.IsNone()) {
+        throw std::runtime_error("Failed to unpack service FS: " + err.mValue);
+    }
+
+    fs::remove_all(imageParts.mServiceFSPath);
+
+    auto [rootFSDigest, error] = aos::common::utils::HashDir(tmpRootFS);
+    if (!error.IsNone()) {
+        throw std::runtime_error("Failed to calculate rootfs checksum: " + rootFSDigest);
+    }
+
+    aos::common::utils::ValidateDigest(rootFSDigest);
+
+    auto [_, hex] = aos::common::utils::ParseDigest(rootFSDigest);
```

Comment:
cpp check might complain on "_"  being an unused variable
Can you please double check if it passes static scan